src/main/java/soot/toolkits/graph/CytronDominanceFrontier.java

```diff
@@ -109,35 +110,28 @@ protected void bottomUpDispatch(DominatorNode<N> node) {
    *              if (idom(Y)!=X) then DF(X) <- DF(X) U Y
    *        end
    *      end
+   * </code>
    * </pre>
-   **/
+   */
   protected void processNode(DominatorNode<N> node) {
-    List<DominatorNode<N>> dominanceFrontier = new ArrayList<DominatorNode<N>>();
+    HashSet<DominatorNode<N>> dominanceFrontier = new HashSet<DominatorNode<N>>();
```

Comment:
Why a set?

Reply:
The Cytron algorithm defines the dominance frontier as an unordered set. The issue with using a List in the implementation is that some nodes may be inserted more than once. Since the tree is traversed bottom-up, using the results from lower nodes in the tree to compute the result for higher nodes, the duplication is especially disastrous when it occurs lower in the tree because it can be duplicated again for each level up the tree thus exponentially increasing the size of the List, with each entry being duplicated many times.

Reply:
I ran into this issue while processing part of Oracle JDK 8 (I don't remember which class/method specifically) and replacing the List with a Set significantly improved performance while still giving the correct outcome in code that uses the dominance frontiers.